pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -184,8 +184,10 @@ func (hsm *hostStateMachine) handleRegistering(info *reconcileInfo) actionResult
 		// eliminate the need to determine which state we came from here.
 		switch {
 		case hsm.Host.Spec.ExternallyProvisioned:
+			info.host.ResetErrorCount()
```

Comment:
I think you've caught all of the steady states and are resetting the error count correctly. We've discussed other changes to the state machine, though, that would add or remove states, and I worry that we will have to think about the error counter for each change and ensure we add this new call at each of the new transitions.

What do you think about adding a `SetNextState()` (or `Transition()`) function on the `hostStateMachine`, and instead of setting `NextState` directly we invoke that function? Then inside the function we can look at the state, and if it's a steady state we can reset the counter. That would mean we reset the counter in one place, like we have one place that increments it.

Are there cases where we enter a state where sometimes we do want to reset the counter and sometimes we do not?

Reply:
Good point. I think that moving it into `ClearError()` as suggested by @zaneb seems a good approach that could also address such concern indirectly (and it will help in keeping the error logic management in one place only).